odoo/addons/base/models/ir_actions_report.py

```diff
@@ -517,10 +517,28 @@ def barcode(self, barcode_type, value, width=600, height=100, humanreadable=0, q
             if barcode_type == 'QR' and quiet:
                 bar_border = 0
 
-            barcode = createBarcodeDrawing(
-                barcode_type, value=value, format='png', width=width, height=height,
-                humanReadable=humanreadable, quiet=quiet, barBorder=bar_border
-            )
+            kwargs = {
+                'value': value,
+                'format': 'png',
+                'width': width,
+                'height': height,
+                'humanReadable': humanreadable,
+                'quiet': quiet,
+                'barBorder': bar_border,
+            }
+
+            # The QR code can have different layouts depending on the Error Correction Level
+            # See: https://en.wikipedia.org/wiki/QR_code#Error_correction
+            # Level 'L' – up to 7% damage   (default)
+            # Level 'M' – up to 15% damage  (i.e. required by l10n_ch QR bill)
+            # Level 'Q' – up to 25% damage
+            # Level 'H' – up to 30% damage
+            if barcode_type == 'QR' and qr_level:
+                if not isinstance(qr_level, str) or qr_level not in ('L', 'M', 'Q', 'H'):
+                    raise ValueError("QR code Error Correction Level is invalid")
```

Comment:
Do we really care, or should this just fallback to doing nothing?

Reply:
Reading the `try/except`, I'm confused about that.
If it's a `Code128`, then an error is raised.
Otherwise, a Code128 is printed... ?

Reply:
Yes the existing fallback of everything to Code128 is odd, but I was mostly wondering if there was really a point to raising an error on an invalid `qr_level`, why not just not set one and get the default? Possibly log a warning but it doesn't seem to warrant the entire thing blowing up.

Reply:
I disagree, it should blow up, here's why:

1) The layout difference between different Error Correction Level is very hard to be seen just by looking at the barcode, so you won't notice you have a wrong QR code. The warning wouldn't be visible at all to the user, it will just be a line in the log that no one sees. 

2) This error will raise only if someone does give an explicit incorrect `qr_level`. If someone writes `"a"` in `humanreadable` then it's ValueError, and it goes to the `except` block, then it calls `barcode` recursively with the same `humanreadable` value, goes to the `except` again, and raises `ValueError`. I actually see no reason to handle `qr_level` differently.

Reply:
> The layout difference between different Error Correction Level is very hard to be seen just by looking at the barcode, so you won't notice you have a wrong QR code. The warning wouldn't be visible at all to the user, it will just be a line in the log that no one sees.

In 99.9% of cases this is not at all relevant, the user primarly wants a qrcode and the "wrong" level of ECC won't usually matter.

> This error will raise only if someone does give an explicit incorrect qr_level. If someone writes "a" in humanreadable then it's ValueError, and it goes to the except block, then it calls barcode recursively with the same humanreadable value, goes to the except again, and raises ValueError. I actually see no reason to handle qr_level differently.

Instead of getting the qrcode with an ECC different than what was requested they'll be getting a non-qrcode, that seems like an utterly idiotic failure mode.

Reply:
> In 99.9% of cases this is not at all relevant
Well, in the case we are actually making this change for (which makes up for the 100% of the real cases), it's legal requirement - it causes the bill to be refused. Otherwise, you wouldn't specify it.

>Instead of getting the qrcode with an ECC different than what was requested they'll be getting a non-qrcode
Let's fix the `except` block instead and make it `raise` properly. It makes absolutely no sense that you get a Code128 fallback in case of error on a QR code - it's a bug in itself.

If you call an API with explicitly wrong parameters, it raises with a proper Value/AttributeError. You fix it the first time you use it and you're ok with the next.